lang/java/mapred/pom.xml

```diff
@@ -189,7 +189,11 @@
         <artifactId>snappy-java</artifactId>
         <scope>test</scope>
     </dependency>
-
+    <dependency>
+      <groupId>com.github.luben</groupId>
+      <artifactId>zstd-jni</artifactId>
+      <scope>test</scope>
+    </dependency>
```

Comment:
I don't see why the mapred module needs this.

Reply:
TestAvroKeyOutputFormat.java line 87.  This calls CodecFactory.zstandardCodec, which eventually causes the construction of the codec.  The codec validates its input against `Zstd.maxCompressionLevel()` and therefore needs to locate that class at that time.

This more eagerly finds the missing dependency.  I can make the input validation lazy, and do it inside `compress()` instead, which is where we would get the ClassNotFoundError before this change.

Either way, the user gets a ClassNotFoundError if they actually try and compress a block.

Reply:
That's the point....  Nothing in the mapred module causes any sort of compression to be enabled/triggered.   Thus, it should NOT require the zstd-jni library to be on the class path.    This is working without your change.

Reply:
Essentially, its the same reason that snappy is requried in test scope a few lines higher in the POM.

We could try and make them all equally lazy (only get ClassNotFoundError when calling compress() but not in constructor), or just not worry about it.  If you want to use it, you have to have it in your classpath.

Reply:
Snappy is different... There is a test in mapred that specifically checks for and uses snappy.  Thus, it has to be there.      What's more, if it's not there, you don't get a "ClassNotFoundException" (or similar), it's a specific "AvroRuntimeException: Unrecognized codec: snappy".      Basically, if snappy is not on the class path, nothing happens unless you actually try to use it.      Most likely, just updating the "CodecFactory zstandardCodec()" method in CodecFactory to catch all the exceptions and return null like snappy does would "work", but I kind of consider that crappy as there isn't anything logged or anything as to WHY the snappy codec isn't registered.

Reply:
Should we create a different JIRA for cleaning up the codecs and making it consistent?  One thing about ClassNotFoundError or other linkage errors is that the problem is usually brutally obvious.  Though a message that indicates that Avro expects ABC on the classpath to use codec XYZ would be nice.

Reply:
I'll push a version that makes the failure lazier (when you call compress() as before), removes the additional test dependency in mapred, and adds an option to enable / disable the checksum feature.

Reply:
Hmm, interestingly mapred still needs the test dependency, whether the level validation is in the constructor or not, since creating an instance loads the class and the class references the ZStandard code.  That can be lifted out into another lazy inner class, but I'm not sure its worth it.

Reply:
re: snappy, "Unrecognized codec: snappy" is a bad message, it is recognized, but not available.  It should say something more like "Unvailable codec: snappy.  Snappy libraries not found on classpath"